nep-16.mediawiki

```diff
@@ -0,0 +1,109 @@
+<pre>
+  NEP: 16
+  Title: NEO Executable Format (NEF)
+  Author: Fernando Díaz Toledano <shargon@gmail.com>, Erik Zhang <erik@neo.org>, Igor Machado Coelho <igormachado@gmail.com>, Li Jianying <lightsever@hotmail.com>, Vitor Nazário Coelho <vncoelho@gmail.com>
+  Type: Standard
+  Status: Accepted
+  Created: 2020-03-31
+</pre>
+
+==Abstract==
+
+Compiled code to be executed by the Neo virtual machine is represented using a binary format, typically stored in a file, known as the <code>nef</code> file format. The <code>nef</code> file contains Neo bytecodes, as well as other ancillary information such as byte ordering and compiler used.
+
+This NEP describes the NEF standard for NEO smart contracts.
+
+==Motivation==
+
+For the sake of security, we need to place strong syntactic and structural constraints on the executable file. With the <code>nef</code> file format, we can guarantee both the validity and functionality of the compiled file.
+
+==Specification==
+
+In the field definitions below, we provide both the definitions of the fields as well as the corresponding functions.
+
+==NefFile Structure==
+
+A <code>nef</code> file consists of a stream of 8-bit bytes. Multibyte data items are always stored in little-endian order, where the low bytes come first. Its structure is described further below.
+
+<pre>
+nef_file
+{
+  4-byte  magic;
+  32-byte compiler;
+  32-byte version;
```

Comment:
Since they are both `string`, should we combine them to a single field? @neo-project/core What do you think?

Reply:
I think that would be ok, @erikzhang, because a compiler and version represent a single entity.

Reply:
> Since they are both string, should we combine them to a single field? @neo-project/core What do you think?

I agree Erik, because version only makes sense together with compiler information. Maybe it's interesting to indicate some desired standard. Example: `compilerName-v0.0.0`. Or maybe, `compilerName-lang-v0.0.0`. Official could be something like `core-cs-v0.0.0`, or any name you prefer. This allows to easily understand origin of contract, like `boa-py-v0.0.0`.